### ISSUE TITLE
Should `rectangular` mean non-sparse for a domain?

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#19268]
This issue asks whether the concept `rectangular` domains should include `sparse` domains.

This decision affects the [Domains chapter of the language spec](https://chapel-lang.org/docs/master/language/spec/domains.html) and the predicates [`domain.isRectangular()`](https://chapel-lang.org/docs/master/language/spec/domains.html#ChapelDomain.isRectangular) and [`array.isRectangular()`](https://chapel-lang.org/docs/master/language/spec/arrays.html#ChapelArray.isRectangular) . The documentation needs to be clear about this topic.

## Background

The current language specification defines [here](https://chapel-lang.org/docs/master/language/spec/domains.html#domain-overview) the following concepts:

* A domain is exactly one of the following:
  - base rectangular domain
  - simple rectangular subdomain
  - sparse rectangular subdomain
  - base associative domain
  - simple associative subdomain
  - sparse associative subdomain (currently not implemented)

* Broader domain categories are:
  - regular: a compact representation of a rectangular index space
    i.e. rectangular non-sparse domains+subdomains
  - irregular: all others
    i.e. associative domains+subdomains and sparse subdomains

In practice we typically use a simpler classification instead, where a domain is exactly one of:
  - rectangular (excludes sparse, so the same as "regular")
  - associative (excludes sparse)
  - sparse (includes only sparse rectangular, as we do not have sparse associative)

Associative and sparse domains are collectively called "irregular".

This simpler classification is reflected in the current behavior of the methods `.isRectangular()` `.isAssociative()` `.isSparse()` `.isIrregular()` on domains and arrays.

## Proposal

This issue proposes to exclude `sparse` subdomains from the category of `rectangular` domains. Then, eliminate the `regular` category as an unnecessary synonim for "rectangular".

This proposal takes the view that a sparse rectangular domain is not really rectangular.

The key benefit of this proposal is a simpler, clearer classification of domains -- the one we typically use in practice, see above.

If the user wants to distinguish between sparse subdomains of rectangular vs. associative domains when we have the latter, they can query the sparse domain's parent. We can also define the methods `.isSparseRectangular()` and `isSparseAssociative()` if desired.

In today's domain module re-review, we liked this proposal, however we also see the merit in the current classification. Therefore...

## Counter-proposal

DO include sparse rectangular subdomains in the `rectangular` category. This reflects the view that sparse rectangular subdomains DO define a rectangular iteration space, in their own way. Also, this follows the lead of Matlab.

The main challenge with this counter-proposal is TO NAME the category of rectangular non-sparse. The name we have for it currently is `regular`, which is not intuitive.

**Comment by bradcray** [comment]
I'm OK with the status quo here.  For sparse rectangular domains, I can see `isRectangular()` meaning either `false` (because the explicitly stored indices are not guaranteed to be rectangular in shape) or `true` (because the conceptual index set described by the domain is rectangular — i.e., something within the rectangular bounding box is not "out of bounds" for the array, it's just not explicitly represented).  But I would probably lean more towards `false` because (a) it's the status quo, so means fewer changes/less work (and nobody has ever complained about it that I can recall, though I personally have to check what decision we made each time) and (b) I think this is the common case to query for (it's rare—though not unthinkable—to want to say "Tell me if this is either sparse or dense rectangular"), so it's nice to keep it short and sweet.

If someone (like @stonea) wanted to advocate for `true` and garner support for it on the team, I'd still be open to that and would propose that we use `isDenseRectangular()` in place of `isRectangular()` today.

In either case, I'd be inclined to deprecate `isRegular()`.  I think `isIrregular()` and `!isIrregular()` are clearer and more self-descriptive than `isRegular()`.  If `isRegular()` ends up being redundant with `isRectangular()` then it's silly to have both.  If we need it to be different, I'd prefer `isDenseRectangular()`.

**Comment by bradcray** [comment]
> How do strided domains fit into this? They're also (as of today) considered rectangular right?

RIght.

> Why do we treat strided domains as being their own kind of base domain but then treat sparse as being a subdomain refining some other dense domain?

Sparse domains need a parent domain in order to distinguish between "in-bounds indices that correspond to zero/IRV values" (e.g., (2,3) from {1..n, 1..n} where (2,3) is not one of the indices explicitly stored by the domain) vs. "out-of-bound indices" (e.g., (-5, -5) in a sparse subdomain over {1..n, 1..n}).  Strided domains don't have a notion of "in-bounds but not represented" like sparse domains do which is why only sparse domains don't require a distinct base domain today.  E.g., `{1..10 by 2}` says everything I need to about the index set, whereas the sparse index set `{(1,1), (3, 5), (7,9)}` doesn't tell me what the parent domain is.

That said, if we invested more in subdomains (as we originally intended to, and may yet return to), it would be accurate to say for a code like:  `const D = {1..n};  constD2 = D by 2;` that `D2` is a (dense) subdomain of `D`.  Similarly, when we do domain slicing like `D3[D4]` we've tended to think of the result as being a subdomain of D3.  However, we don't do much with this today beyond maybe use it as terminology.

**Comment by bradcray** [comment]
> If I think in terms of sparse arrays (rather than domains) then it seems to be that a sparse array should be considered rectangular on the premise that its interface presents itself as such (just returning IRV for the non sparse values).

I tend to agree that "isRectangular" on a sparse array feels more like `true` and on a domain feels more like `false`.  But I think we need to pick the same for both.  If I were arguing the array perspective towards `false`, I might say "Its implementation doesn't support 'math' to compute an offset from a pointer given its indices".  If I were arguing the domain perspective towards `true`, I might say "it represents all of the indices in the bounding box, but only some of them explicitly."

Which brings up an interesting question:  When we do `mySparseDomain.contains((i,j))` what question is it answering?  "Yes this index is within my bounding box" or "This is one of the explicitly represented indices"?

I'm noting that @lydia-duncan put the Chapel 2.0 label on this, but I think it could be argued that if we're calling sparse "unstable" for Chapel 2.0, this question doesn't technically need to be answered in that timeframe (in that we won't change the answer for dense arrays/domains in any case, and since sparse is unstable, it could be reasonable for it to change).  Or am I missing some other subtlety?

**Comment by stonea** [comment]
> Sparse domains need a parent domain in order to distinguish between "in-bounds indices that correspond to zero/IRV values" (e.g., (2,3) from {1..n, 1..n} where (2,3) is not one of the indices explicitly stored by the domain) vs. "out-of-bound indices" (e.g., (-5, -5) in a sparse subdomain over {1..n, 1..n}). Strided domains don't have a notion of "in-bounds but not represented" like sparse domains do which is why only sparse domains don't require a distinct base domain today. E.g., {1..10 by 2} says everything I need to about the index set, whereas the sparse index set {(1,1), (3, 5), (7,9)} doesn't tell me what the parent domain is.

Just to play devil's advocate (I think we're well past the point where we'd want to change what we consider a base domain vs. a subdomain, but maybe this conversation will help us figure out what's confusing and figure out how to better document\explain it):

Don't both sparse and strided domains have to be defined as something that's refining some kind of rectangular domain?  I suppose you could say 'no': strided domains are defined in terms of strided ranges not domains but isn't a strided range just a kind of 1-dimensional domain?

Sure, in the strided array accesses to a "non strided over" element results in (I assume a runtime) error whereas in the sparse case it returns IRV, but doesn't that says more about the behavior that occurs when you apply the domain to an array.

> E.g., {1..10 by 2} says everything I need to about the index set, whereas the sparse index set {(1,1), (3, 5), (7,9)} doesn't tell me what the parent domain is.

Isn't this just pointing out that (as of today) we lack syntax to define a sparse literal?  Sure that syntax would have to define it in terms of a sparse index set being applied to some other domain, but is that really that much different than the 'by' operator being applied to the range?

I could imagine something like this:

```chpl
var mySparseDom = {[(1,1), (3,5), (7,9)] in {1..10, 1..10})
```

Also, just as a curiosity: can you have sparse subdomains on strided domains? If so does it error out or give you the IRV when you access the non strided value?

> I tend to agree that "isRectangular" on a sparse array feels more like true and on a domain feels more like false. But I think we need to pick the same for both.

Yeah I agree that whatever behavior we decide is right should be consistent across arrays and domains. It'd be super confusing otherwise.

**Comment by bradcray** [comment]
> I suppose you could say 'no': strided domains are defined in terms of strided ranges not domains

I would say that, yeah.  Specifically, given the program `var D = {1..n by 2, 1..n by 2}`, today, the compiler will never create a domain `{1..n, 1..n}`, nor do I see any reason that it would need to.  So I don't think of `D` as being a subdomain.

If instead we wrote:  `var D = {1..n, 1..n}, D2 = D by (2,2);` then I'd be more inclined to view D2 as a subdomain of D, and if you wrote: `var D = {1..n, 1..n}, D2: subdomain(D) = D by (2,2);` then I definitely would consider it to be.

> but isn't a strided range just a kind of 1-dimensional domain?

I'd say "no", though I agree they're obviously similar, much as an int(8) is similar to, yet different than, an uint(64) or (maybe better) a record containing just a single int(8) field.

> but doesn't that says more about the behavior that occurs when you apply the domain to an array then something about the domain itself?

When you say "apply the domain to an array" do you mean "declare an array over the domain?" (as opposed to "slice into the array with the domain," say?).  If so, then yes.  All arrays are declared over domains, and the nature of the domain (local vs. distributed, block vs. cyclic, rectangular vs. associative, etc.) governs the implementation and behavior of the array.

> Isn't this just pointing out that (as of today) we lack syntax to define a sparse literal? Sure that syntax would have to define it in terms of a sparse index set being applied to some other domain, but is that really that much different than the 'by' operator being applied to the range?

I'm not seeing where you're going with this, but: I think it's accurate that we don't currently have a syntax for creating a sparse domain literal and that it would be nice to either have one, or to have an expression that would create one.  For example, if I wanted to express a sparse subdomain without a declared type, it might feel natural to do so by slicing a dense domain with an associative domain with matching idxType, like: `{1..n, 1..n}[{(1, 1), (2, 5), (3, 7)}]`. But I wouldn't call that a literal so much as an operation whose result is a sparse subdomain (and I don't think this is implemented today, so it's just a speculative idea).  Or we could pick another form like yours. 

> can you have sparse subdomains on strided domains?


Yes.

> If so does it error out or give you the IRV when you access the non strided value?

You get an OOB error:  https://tio.run/##S85ILEjN@f8/OT8vLTNdAUgVlyjkKdgqGBpYc5UlFim4ANnVhnp6eQpJlQpGOgpwZi1UPthKobggsag4VaG4NCklPzcxM0/DRRMi6WilEO0SHKtQlJqYY81VXpRZkpqTp@EYbahjGKuJLGCkYwQU@P8fAA

I'm realizing that I never got to the last question of your previous comment, though in part it's because I'm not 100% sure. 
 Taking a stab at it:

> I think right now what we have is this right?
> **Kinds of domains and subdomains:**
> - Regular
>   - Rectangular
>     - Strided
>     - Non-strided
> - Irregular (also non rectangular)
>   - Associative
>   - Sparse (always a subdomain)

Perhaps... (Vass would probably know best at this point).  I would've guessed that Regular and Rectangular would've been on a single line like `-Regular == Rectangular`.


> And this issue is proposing to change it to this:
> 
> **Kinds of domains and subdomains:**
> - Rectangular
>   - Strided
>   - Non-strided
> - Non rectangular
>   - Associative
>  - Sparse (always a subdomain)

The main part of both these summaries that I balk at is that I consider sparse to be more like another dimension that intersects the 'rectangular | associative' dimension rather than part of a well-nested taxonomy.  So I might change your hierarchy into something more like the following (if you can envision this ugliness as a 3x2 table):

```
                                dense  |    sparse
             non-strided
            /
rectangular
            \
            strided

-----

associative
```

That said, even that's not particularly satisfying because it feels like a laziness or oversimplification that a rectangular must always, as a whole, be strided or non-strided rather than (ultimately, someday, maybe) supporting a mixture of stridednesses in different dimensions, like `{1..n, 1..10 by 2}`.  In which case it feels simpler to me to say "{rectangular | associative} x {sparse | dense}`, where rectangular domains can be defined by strided or unstrided ranges (where, after all, an unstrided range is arguably just a strided range whose stride must always be 1?).

(I actually continue to think that it's probably fine if we never mix stridednesses within a single domain in practice, but if I were defining a formal taxonomy, I probably wouldn't paint myself into a corner on it).

**Comment by lydia-duncan** [comment]
> I'm noting that @lydia-duncan put the Chapel 2.0 label on this, but I think it could be argued that if we're calling sparse "unstable" for Chapel 2.0, this question doesn't technically need to be answered in that timeframe (in that we won't change the answer for dense arrays/domains in any case, and since sparse is unstable, it could be reasonable for it to change). Or am I missing some other subtlety?

I might have been a bit overzealous, but I think it would be reasonable to have the label to track that we need to mark the call on sparse arrays and domains as unstable before 2.0

**Comment by bradcray** [comment]
> to track that we need to mark the call on sparse arrays and domains as unstable before 2.0

I don't think we need to (or should) do that.  Specifically, once we label any use of `sparse` as unstable (https://github.com/Cray/chapel-private/issues/2893), I don't think we will also need to mark every call that interacts with a sparse array/domain as also being unstable (because then anyone doing sparse programming—which is still reasonable to use and give us feedback on—will get us tons of warnings all delivering the same essential message: "sparse is still very much in flux").  So I think just a blanket "We see you're using sparse; note that it's still a work-in-progress and therefore unstable" from the original sparse domain declaration should be sufficient (where I'm specifically imagining that the parser might just print this out whenever it sees the `sparse` keyword).

**Comment by lydia-duncan** [comment]
Ah, good point!  I've removed the label

**Comment by stonea** [comment]
As a side note:

This sentence is in the [doc](https://chapel-lang.org/docs/main/language/spec/domains.html): "Rectangular domains, with the exception of sparse subdomains, are regular." today (which would seem to imply that sparse subdomains can be rectangular).

Anyway if we decide that rectangular means non-sparse we should remove that sentence.

Also I was playing around with all this and found it a helpful exercise to make this table to see what the behavior is today (maybe its something else we'd want to put in the doc):

```
                              isRectangular()   isRegular()   isSparse()
rectangular domain            true              true          false
associative domain            false             false         false
simp. subdom of rect dom      true              true          false
simp. subdom of assoc dom     false             false         false
sparse subdom of rect dom     false             false         true 
sparse subdom of assoc dom    --- this combination does not exist ---
sparse subdom of sparse sbdm  false             false         true 
```
